src/spikeinterface/core/generate.py

```diff
@@ -230,6 +230,78 @@ def add_synchrony_to_sorting(sorting, sync_event_ratio=0.3, seed=None):
     return sorting
 
 
+def generate_injected_sorting(
+    sorting: BaseSorting,
+    num_samples: List[int],
+    max_injected_per_unit: int = 1000,
+    injected_rate: float = 0.05,
+    refractory_period_ms: float = 1.5,
+) -> NumpySorting:
+    injected_spike_trains = [{} for seg_index in range(sorting.get_num_segments())]
+    t_r = int(round(refractory_period_ms * sorting.get_sampling_frequency() * 1e-3))
+
+    for segment_index in range(sorting.get_num_segments()):
+        for unit_id in sorting.unit_ids:
+            spike_train = sorting.get_unit_spike_train(unit_id, segment_index=segment_index)
+            n_injection = min(max_injected_per_unit, int(round(injected_rate * len(spike_train))))
+            # Inject more, then take out all that violate the refractory period.
+            n = int(n_injection + 10 * np.sqrt(n_injection))
+            injected_spike_train = np.sort(
+                np.random.uniform(low=0, high=num_samples[segment_index], size=n).astype(np.int64)
+            )
+
+            # Remove spikes that are in the refractory period.
+            violations = np.where(np.diff(injected_spike_train) < t_r)[0]
+            injected_spike_train = np.delete(injected_spike_train, violations)
+
+            # Remove spikes that violate the refractory period of the real spikes.
+            # TODO: Need a better & faster way than this.
+            min_diff = np.min(np.abs(injected_spike_train[:, None] - spike_train[None, :]), axis=1)
+            violations = min_diff < t_r
+            injected_spike_train = injected_spike_train[~violations]
+
+            if len(injected_spike_train) > n_injection:
+                injected_spike_train = np.sort(np.random.choice(injected_spike_train, n_injection, replace=False))
+
+            injected_spike_trains[segment_index][unit_id] = injected_spike_train
+
+    return NumpySorting.from_unit_dict(injected_spike_trains, sorting.get_sampling_frequency())
+
+
+class TransformedSorting(BaseSorting):
+    def __init__(self, sorting, added_spikes=None, refractory_period_ms=5):
+        fs = sorting.get_sampling_frequency()
+        unit_ids = sorting.get_unit_ids()
+
+        BaseSorting.__init__(self, fs, unit_ids)
+        sorting.precompute_spike_trains()
+        assert added_spikes.dtype == minimum_spike_dtype, "added_spikes should be a spike vector"
+
+        self._cached_spike_vector = np.concatenate((sorting._cached_spike_vector, added_spikes))
+        self.added = np.zeros(len(self._cached_spike_vector), dtype=bool)
+        self.added[len(sorting._cached_spike_vector) :] = True
```

Comment:
wait shouldn't `self.added` correspond to the indices of the added spikes?

Reply:
Could be, I thought added was a boolean mask with True or False for added spikes. But I can save only the indices, we just need to decide

Reply:
Ahhh sorry I misread the logic. You first concatenate and then set the last Len(added_spikes) to True! Looks good ;)